### ISSUE TITLE
Auto-indentation for lists

### ISSUE THREAD

**Issue by telemachus** [preservim/vim-markdown#126]
Hi,

I believe that my problem is related to #78 and I'm pretty sure that this what I'm seeing is a deliberate choice (based on lines 41-43 of `indent/mkd.vim`), but I want to make sure.

If I'm writing Markdown and I start a list-item, when I hit return, the cursor is indented one level _further_ in than the original list-item.

Example:

```
Here is some text.

+ Now I start a list. Soon I will hit enter.<CR>
    |<cursor is now here>
```

I don't understand this choice. As a user, when I'm done with the first list-item (and hit `<CR>`), I expect to start typing the next list-item. So shouldn't I be at the same indentation level, rather than further in? (If I want the next list-item to start a sub-list or be code, that's a separate choice. I would then `<TAB>` in further myself.)

tl;dr Is this observed indentation behavior deliberate? If so, why? Thanks.

**Comment by mauvm** [comment]
```
  formatlistpat=^\s*\d\+[\]:.)}\t ]\s*
  formatoptions=croql
        Last set from /usr/share/vim/vim73/ftplugin/vim.vim
  filetype=vim
        Last set from /usr/share/vim/vim73/filetype.vim
  indentexpr=GetVimIndent()
        Last set from /usr/share/vim/vim73/indent/vim.vim
  comments=sO:" -,mO:"  ,eO:"",:"
        Last set from /usr/share/vim/vim73/ftplugin/vim.vim
  textwidth=78
        Last set from /usr/share/vim/vim73/ftplugin/vim.vim
```

`set fo` gives me `formatoptions=rtcqln`

**Comment by MWers** [comment]
:+1: 

```
  formatlistpat=^\s*\d\+\.\s\+\|^[-*+]\s\+
        Last set from /usr/share/vim/vim73/ftplugin/markdown.vim
  formatoptions=rtcqln
        Last set from /usr/share/vim/vim73/ftplugin/markdown.vim
  filetype=markdown
        Last set from ~/.vimrc
  indentexpr=GetMkdIndent()
        Last set from ~/Dropbox/Application Support/Dotfiles/.vim/bundle/vim-markdown/indent/mkd.vim
  comments=fb:*,fb:-,fb:+,n:>
        Last set from /usr/share/vim/vim73/ftplugin/markdown.vim
  textwidth=0
```

**Comment by sukima** [comment]
This was referenced above but I don't see a reply.

List indent is mentioned in the [original spec](http://daringfireball.net/projects/markdown/syntax):

> List items may consist of multiple paragraphs. Each subsequent paragraph in a list item must be indented by either 4 spaces or one tab

This allows a markdown parser to bundle multiple paragraphs in one list item. By aligning the text to follow suit you can make the switch to double digit numbers without offsetting the indent as shown:

```markdown
8. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
   tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
9. quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
   Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat
10. nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui
    officia deserunt mollit anim id est laborum. 
```

VS

```markdown
8.  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
    tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
9.  quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
    Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat
10. nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui
    officia deserunt mollit anim id est laborum. 
```

**Comment by fredsherbet** [comment]
I'm hitting this issue, and found that it's because I remapped `<CR>` to `<esc>o` (i.e. whereever I am in the line, return starts a new line, rather than splitting the line)

When I removed the mapping, I get the correct list behaviour.

Does that give any clue to the problem? I'd quite like to keep my remapped return key, too...

**Comment by fredsherbet** [comment]
I worked around by using

    inoremap <CR> <esc>A<CR>

**Comment by tdi** [comment]
```
formatlistpat=^\s*\d\+\.\s\+\|^[-*+]\s\+
    Last set from /usr/share/vim/vim74/ftplugin/markdown.vim
formatoptions=tcqln
    Last set from /usr/share/vim/vim74/ftplugin/markdown.vim
filetype=mkd.markdown
    Last set from ~/.vim/bundle/vim-markdown/ftdetect/markdown.vim
indentexpr=GetMarkdownIndent()
    Last set from ~/.vim/bundle/vim-markdown/indent/markdown.vim
comments=fb:*,fb:-,fb:+,n:>
    Last set from /usr/share/vim/vim74/ftplugin/markdown.vim
textwidth=100
    Last set from ~/.vimrc
```

**Comment by mbsmithlaw** [comment]
I also am having this issue.  

``` 
formatlistpat=^\s*\d\+\.\s\+\|^[-*+]\s\+
        Last set from /usr/local/Cellar/macvim/7.4-77/MacVim.app/Contents/Resources/vim/runtime/ftplugin
/markdown.vim
  formatoptions=tcqln
        Last set from /usr/local/Cellar/macvim/7.4-77/MacVim.app/Contents/Resources/vim/runtime/ftplugin
/markdown.vim
  filetype=markdown.pandoc
  indentexpr=GetMarkdownIndent()
        Last set from ~/.vim/bundle/vim-markdown/indent/markdown.vim
  comments=fb:*,fb:-,fb:+,n:>
        Last set from /usr/local/Cellar/macvim/7.4-77/MacVim.app/Contents/Resources/vim/runtime/ftplugin
/markdown.vim
  textwidth=0 
```

**Comment by arthurdent** [comment]
I moved some things around in my vimrc and now things are working correctly.

`filetype plugin indent on` needs to be above `syntax enable` or `syntax on`

Here is the top of my working vimrc:
```
execute pathogen#infect()
filetype plugin indent on
syntax enable
```

Removing `filetype plugin indent on` from my vimrc fixed the problem, but that wasn't a good solution.

Alternatively, I was able to fix it by manually issuing `:syntax enable` in vim. I thought this was weird because it was already in my vimrc, but eventually I put two and two together.

**Comment by sukima** [comment]
@arthurdent I think this might be relevant: https://github.com/VundleVim/Vundle.vim/issues/176

It is Vundle not pathogen but I think I recall them both having the same gotcha.

P.S. :+1: on the GitHub username.

**Comment by danawoodman** [comment]
@arthurdent moving `filetype plugin indent on` did not work for me. Are you sure that was the only change you made?

FWIW, my `:verb set flp fo ft inde comments tw` output:

```
  formatlistpat=^\s*\d\+[\]:.)}\t ]\s*
        Last set from ~/.vim/view/~=+.vimrc=
  formatoptions=jcroql
        Last set from ~/.vim/view/~=+.vimrc=
  filetype=vim
        Last set from /usr/local/Cellar/neovim/HEAD/share/nvim/runtime/filetype.vim
  indentexpr=GetVimIndent()
        Last set from ~/.vim/view/~=+.vimrc=
  comments=sO:" -,mO:"  ,eO:"",:"
        Last set from ~/.vim/view/~=+.vimrc=
  textwidth=500
        Last set from ~/.vim/view/~=+.vimrc=
```

**Comment by borismus** [comment]
Also seeing this. What's the solution?

**Comment by arthurdent** [comment]
@danawoodman The only other thing I did is disable folding with `let g:vim_markdown_folding_disabled=1`, which I also added before `syntax enable`. I should definitely mention that using `o` or `O` to enter insert mode still results in an indent instead of a new item. Only pressing return in insert mode adds an item. I am not sure if this is the intended behavior.

**Comment by flight16** [comment]
I'm also getting this.  The only solution for me was to remove "filetype plugin indent on" from vimrc. : (

This is really, really annoying.  I have stayed away from this plugin for years because of this behavior.

I'm using Vim 7.4.258 on OS X.

**Comment by flight16** [comment]
As a work-around to only disable plugin indents with markdown, I added the following to my .vimrc: 

    autocmd BufNewFile,BufRead *.md filetype plugin indent off

With the number of people posting in this thread, maybe it should go in the README.md so people find it easier?

**Comment by telemachus** [comment]
@shirosaki An alternative is to switch to [tpope's markdown plugin](https://github.com/tpope/vim-markdown).

**Comment by j5shi** [comment]
@blueyed I'm having a similar issue, but in my case both `*` and `+` work well, except `-`, which will get over indented by 2 spaces.

```viml

  formatlistpat=^\s*[0-9*-+]\+[\]:.)}\t ]\s*\|^\s*\[[x\- ]\]\s\+\|^\s*\[[x -]\]\s\+
	Last set from c:\Program Files (x86)\Vim\_vimrc
  formatoptions=Bjtqlnr
	Last set from c:\Program Files (x86)\Vim\vimfiles\bundle\vim-markdown\indent\markdown.vim
  filetype=markdown
	Last set from c:\Program Files (x86)\Vim\vimfiles\bundle\vim-markdown\ftdetect\markdown.vim
  indentexpr=GetMarkdownIndent()
	Last set from c:\Program Files (x86)\Vim\vimfiles\bundle\vim-markdown\indent\markdown.vim
  comments=b:>
	Last set from c:\Program Files (x86)\Vim\_vimrc
  textwidth=100
	Last set from c:\Program Files (x86)\Vim\_vimrc
```

**Comment by djhaskin987** [comment]
Still a problem in 2017. I'm on a mac, I just cloned this into my bundle directory. I'm using pathogen.

**Comment by for-coursera** [comment]
How about adding something like that to `after/ftplugin/markdown.vim`

    " helper to detect if you are in a dash-list or not
    " because it's a bug: https://github.com/plasticboy/vim-markdown/issues/126
    function! s:InAListOrNot()
        let the_line = getline(".")
        return the_line =~ '^ *-' ? 1 : 0
    endfunction
    " then map i_CR and n_o to i_CR_BS and n_o_BS accordingly if yes, and leave as is if not
    inoremap <expr> <CR> <SID>InAListOrNot() ? '<CR><BS>' : '<CR>'
    nnoremap <expr> o <SID>InAListOrNot() ? 'o<BS>' : 'o'

**Comment by gpanders** [comment]
I use this plugin indirectly via vim-polyglot and am seeing the same problem. The solution for me was to clear `indentexpr`:
```vim
setlocal indentexpr=
```

**Comment by akarzim** [comment]
Same here using vim-polyglot too.  My `:verb set flp fo ft inde comments tw` output was:

```vim
  formatlistpat=^\s*\d\+\.\s\+\|^[-*+]\s\+\|^\[^\ze[^\]]\+\]:
        Modifié la dernière fois dans /usr/local/Cellar/neovim/0.4.3/share/nvim/runtime/ftplugin/markdown.vim line 14
  formatoptions=jtcqln
        Modifié la dernière fois dans /usr/local/Cellar/neovim/0.4.3/share/nvim/runtime/ftplugin/markdown.vim line 13
  filetype=markdown
        Modifié la dernière fois dans ~/.local/share/nvim/plugged/vim-polyglot/ftdetect/polyglot.vim line 379
  indentexpr=GetMarkdownIndent()
        Modifié la dernière fois dans ~/.local/share/nvim/plugged/vim-polyglot/indent/markdown.vim line 6
  comments=fb:*,fb:-,fb:+,n:>
        Modifié la dernière fois dans /usr/local/Cellar/neovim/0.4.3/share/nvim/runtime/ftplugin/markdown.vim line 12
  textwidth=80
        Modifié la dernière fois dans ~/git/vim/after/ftplugin/markdown.vim line 2
```

And my workaround is to set this conf in my `after/ftplugin/markdown.vim` file: 

```vim
" r Automatically insert the current comment leader after hitting
"   <Enter> in Insert mode.
" o Automatically insert the current comment leader after hitting 'o' or
"   'O' in Normal mode.
set formatoptions+=ro

" The 'comments' option is a comma-separated list of parts.
" Each part defines a type of comment string.
" A part consists of: {flags}:{string}

" {string} is the literal text that must appear.

" {flags}:
"   n Nested comment.  Nesting with mixed parts is allowed.
"   b Blank (<Space>, <Tab> or <EOL>) required after {string}.
set comments=b:*,b:-,b:+,b:1.,n:>
```

Note about the removed `f` flag:

```vim
  f     Only the first line has the comment string.  Do not repeat comment on
        the next line, but preserve indentation (e.g., a bullet-list).
```

bonus point: thanks to `b:1.` it handles numeric lists well too ;)

**Comment by pwc3** [comment]
Based on @akarzim's post above, I was able to work around this issue by adding the following to my `.vimrc`

```vim
au filetype markdown set formatoptions+=ro
au filetype markdown set comments=b:*,b:-,b:+,b:1.,n:>
```

**Comment by bendman** [comment]
The proposed [solution](https://github.com/plasticboy/vim-markdown/issues/126#issuecomment-640890790) from @pwc3 fixed my list indentation both for `return` and `o`/`O`.

**Comment by jchook** [comment]
The fix from @pwc3 works for me with the cursor on a list item, but not on the empty line *after* the list:

```markdown
- List item 1
- List item 2
- List item 3
█ <normal cursor here, press o>
```

```markdown
- List item 1
- List item 2
- List item 3

    | <insert cursor here now>
```

Output of `verb set flp fo ft inde comments tw`:

```
  formatlistpat=^\s*\d\+\.\s\+\|^[-*+]\s\+\|^\[^\ze[^\]]\+\]:
        Last set from /usr/share/nvim/runtime/ftplugin/markdown.vim line 14
  formatoptions=jtqlnro
        Last set from ~/.dotfiles/editor/nvim/init.vim line 616
  filetype=markdown
        Last set from ~/.dotfiles/editor/nvim/init.vim line 599
  indentexpr=GetMarkdownIndent()
        Last set from ~/.dotfiles/editor/nvim/plugged/vim-polyglot/indent/markdown.vim line 8
  comments=b:*,b:-,b:+,b:1.,n:>
        Last set from ~/.dotfiles/editor/nvim/init.vim line 617
  textwidth=0
```

**Comment by devstefancho** [comment]
> 

> Using this hack for now:
> 
> ```viml
> autocmd filetype markdown set indentexpr=
> ```

this works for me.
(when pressing `o` then cursor start from same level of upper list level)

**Comment by alerque** [comment]
Re-openening to (eventually) review why people think they still have a problem with this after a supposed fix was merged.

**Comment by jodavaho** [comment]
I still have this issue:

```
- List item 1
- List item 2
- List item 3
| <cursor here, press enter>
```
```
- List item 1
- List item 2
- List item 3

    | <insert cursor here now>
```

And `autocmd BufNewFile,BufRead *.md filetype plugin indent off` at _bottom_ of `.vimrc` solved it.

**Comment by pizdjuk** [comment]
A have the same issue. Fixed with syntax on. But the problem, that the line wrapping still does not work properly (not indenting)

**Comment by Gullumluvl** [comment]
Hi, same issue. 

> Re-openening to (eventually) review why people think they still have a problem with this after a supposed fix was merged.

@alerque, what was the supposed fix? The commit [642730f](https://github.com/preservim/vim-markdown/commit/642730fccce7e78c995389f3c430747d266efa96) on Nov 16, 2015?

For me the solution was to comment out parts of `GetMarkdownIndent` as indicated by @gpanders [above](https://github.com/preservim/vim-markdown/issues/126#issuecomment-485579068).

Only adding the line `autocmd BufNewFile,BufRead *.md filetype plugin indent off` at the end of the .vimrc *did not* work.
